src/main/java/org/gaul/s3proxy/CrossOriginResourceSharing.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2014-2018 Andrew Gaul <andrew@gaul.org>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gaul.s3proxy;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.base.Joiner;
+import com.google.common.base.Splitter;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableSet;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class CrossOriginResourceSharing {
+    private static final String VALUE_SEPARATOR = " ";
+    private static final String HEADER_VALUE_SEPARATOR = ", ";
+    private static final String ALLOW_ANY_HEADER = "*";
+
+    private static final Logger logger = LoggerFactory.getLogger(
+            CrossOriginResourceSharing.class);
+
+    private final String allowedMethodsRaw;
+    private final String allowedHeadersRaw;
+    private final Set<Pattern> allowedOrigins;
+    private final Set<String> allowedMethods;
+    private final Set<String> allowedHeaders;
+
+    protected CrossOriginResourceSharing() {
+        // CORS Allow all
+        this(".+", "GET PUT POST", ALLOW_ANY_HEADER);
+    }
+
+    protected CrossOriginResourceSharing(String allowedOrigins,
+            String allowedMethods, String allowedHeaders) {
```

Comment:
Should these parameters be `Collection<String>` and the caller can parse them from the configuration format?  Consider potential users that configure via Java code instead of config files.

Reply:
done